cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DataTracerFactoryProvider.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.internal.app.preview;
+
+import co.cask.cdap.app.preview.DataTracerFactory;
+import co.cask.cdap.proto.id.ApplicationId;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A class which provides {@link DataTracerFactory} based on the {@link ApplicationId}
+ */
+public final class DataTracerFactoryProvider {
+  private static final DataTracerFactory DEFAULT_FACTORY = new NoopDataTracerFactory();
+  private static final Map<ApplicationId, DataTracerFactory> FACTORY_MAP = new HashMap<>();
```

Comment:
should we use `ConcurrentHashMap`, then we don't need to synchronize the getter and setter